capstone/src/main/java/com/google/sps/servlets/DeletePollOptionServlet.java

```diff
@@ -50,34 +47,35 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     }
   }
 
-  /**
-   * Sets the variables maxVotes and maxVotesId by iterating through the entities.
-   *
-   * @param results queried results
-   * @return maxVotesId the id of option with maximum number of votes
-   */
-  private long getMaxVotedId(PreparedQuery results) {
-    long maxVotedId = 0;
-    int maxVotes = 0;
-    Iterable<Entity> resultsIterable = results.asIterable();
-    // Set maxVotedId to be first one in iterable of Entities
-    if (Iterables.size(resultsIterable) > 0) {
-      maxVotedId = results.asIterable().iterator().next().getKey().getId();
+  private List<Long> getOptionsList(
+      HttpServletRequest request, HttpServletResponse response, DatastoreService datastore)
+      throws IOException {
+    Entity entity = Group.getGroupEntity(request, response, datastore);
+    List<Long> options =
+        (entity.getProperty("options") == null)
+            ? new ArrayList<Long>()
+            : (List<Long>) entity.getProperty("options");
+    return options;
+  }
+
+  private long getMaxVotedOption(
+      List<Long> options, HttpServletResponse response, DatastoreService datastore)
+      throws IOException {
+    if (options.size() == 0) {
+      return 0;
     }
-    /*
-     * Update maxVotes and maxVotedId by iterating through Entities, checking for
-     * max votes
-     */
-    for (Entity entity : resultsIterable) {
-      long id = entity.getKey().getId();
-      List<String> votes = (ArrayList<String>) entity.getProperty("votes");
+    long maxVotedOptionId = options.get(0);
+    int maxVotes = 0;
+    for (Long optionId : options) {
+      Entity optionEntity = ServletHelper.getEntityFromId(response, optionId, datastore, "Option");
+      List<String> votes = (ArrayList<String>) optionEntity.getProperty("votes");
       int numVotes = (votes != null) ? votes.size() : 0;
       if (numVotes > maxVotes) {
         maxVotes = numVotes;
-        maxVotedId = id;
+        maxVotedOptionId = optionId;
       }
     }
```

Comment:
Maybe better for a follow-up CL (or leaving this as a mental TODO which never needs to get done):

Currently, you're having to do all of this logic every time anybody calls doPost. But one nice property of this calculation is it wouldn't change if nobody added or removed any votes or poll options between `getMaxVotedOption` calls on a given poll.

Could you redesign this servlet/database to allow `O(1)` retrieval in cases where no votes were changes and no poll options were added or removed?

Reply:
I think this servlet is only called when the challenge needs to be updated, so this is handled on the javascript side. I check whether the challenge has passed and if it has, then I call this servlet to delete the top poll option and update the challenge